### PR TITLE
Add `_get_func_code/_is_available` virtual functions to custom nodes

### LINKED ISSUE #3916: VisualShaderNodeCustom decouple node instantiation in vertex/fragment/light from ability to inject code in the stages
### Describe the project you are working on

VFX tools and Godot Engine

Specifically, I'm trying to have all the high level functionalities of spatial/canvas/process material as visual shader nodes (triplanar mapping, canvas animation .. etc.)

### Describe the problem or limitation you are having in your project

It is currently not possible to inject shader code in stages where the node is not instanced.

What I want users to do:
- drag ParticleAnimation node to fragment stage

Particle animation is a great example of this because it requires code on both the vertex stage and the fragment stage.
In vertex stage, INSTANCE_CUSTOM (vertex stage only) is read and it's used to compute new UVs. These new UVs are then used in the fragment shader.
This means that the user needs to instance twice the node, once in vertex stage, and another in fragment stage.
This is not good UX for users, especially because this work that I'm doing is meant to simplify the workflow for non-techincal people. Instancing the node twice requires knowledge of how internally vertex/fragment works.

For comparison, this is how it looks in Unity: 
Screenshot from 2022-02-05 10-30-11
This is plugged directly in the fragment stage.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I'm not highly opinionated about the technical implementation for this, as long as the following workflow is possible:

- User creates a visual shader
- user adds "ParticleAnimation" node in fragment stage and can use output UV to plug in a texture sampler
- no other user action required

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Example implementation:
Add a function to VisualShaderNodeCustom `_is_available` (name up for discussion) that tells VisualShaders wheter or not the node should show up in the "add" popup.
`_get_code` gets executed if the node is present in **any** stage.
This way, as a plugin creator, I am able to inject code in vertex stage even if my node is instanced only in the fragment stage.

This is my own idea for the solution to this problem, but I am not picky on the solution as long as it's possible to give the workflow described above to the users.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It is not possible to deliver this UX with the current VisualShaderNodeCustom implementation.

### Is there a reason why this should be core and not an add-on in the asset library?

This will be a base for addons, or built-in high level shader nodes.

### PR DESCRIPTION
**Pull request by Chaosus**
This targets to fix https://github.com/godotengine/godot-proposals/issues/3916 by adding `_get_func_code` and `_is_available` virtual functions to `VisualShaderNodeCustom` to allow generate a code at the top of each visual shader function. Also, reorganize a code for copy/pasting to disallow showing an option to paste where there are no nodes available to paste. 

cc @QbieShay - if you check this functionality that would be great!

### CODE REVIEW

**Comment by QbieShay** [comment]
Hey @Chaosus , thanks a lot for the swift work!
It works! Thank you so much!!

Do you think it can be backported to 3.x as well?

One minor issue i encountered: if i don't return an empty string when implementing the `_get_func_code` it will write `null` in the shader instead of nothing. So perhaps there should be a check both against empty string and null.
Screenshot from 2022-02-08 14-21-16

Screenshot from 2022-02-08 14-20-36

uncommenting the return fixes the issue.

**Comment by QbieShay** [comment]
Tested with this node:

```
@tool
extends VisualShaderNodeCustom
class_name VisualShaderNodeCustomFlipbookAnimation

func _is_available(mode, type):
	if type == VisualShader.TYPE_FRAGMENT && (mode == VisualShader.MODE_CANVAS_ITEM || mode == VisualShader.MODE_SPATIAL):
		return true
	return false 

func _get_name() -> String:
	return "new_script"

func _get_category() -> String:
	return ""

func _get_description() -> String:
	return ""

func _get_return_icon_type() -> int:
	return PORT_TYPE_SCALAR

func _get_input_port_count() -> int:
	return 3

func _get_input_port_name(port) -> String:
	match port:
		0:
			return "hFrames"
		1:
			return "vFrames"
		2: 
			return "Loop"
	return ""
func get_input_port_default_value(port):
	match port:
		0:
			return 1.0
		1:
			return 1.0
		2: 
			return true

func _get_input_port_type(port) -> int:
	match port:
		2: 
			return PORT_TYPE_BOOLEAN
	return PORT_TYPE_SCALAR_INT

func _get_output_port_count() -> int:
	return 1

func _get_output_port_name(port) -> String:
	return "FlipbookUV"

func _get_output_port_type(port) -> int:
	return PORT_TYPE_VECTOR_2D

func _get_global_code(mode):
	return "varying float lifetime;\n"

func _get_func_code(mode, type):
	match type:
		VisualShader.TYPE_VERTEX:
			return "lifetime = INSTANCE_CUSTOM.y/INSTANCE_CUSTOM.w;\n"
#	return ""

func _get_code(input_vars: Array[String], output_vars: Array[String], mode: Shader.Mode, type: VisualShader.Type) -> String:
	if type == VisualShader.TYPE_FRAGMENT:
		return """
	vec2 flipbook_uv = UV;
	float h_frames = float("""+input_vars[0]+""");
	float v_frames = float("""+input_vars[1]+""");
	float particle_total_frames = float("""+input_vars[0]+""" * """+input_vars[1]+""");
	float particle_frame = floor(lifetime * float(particle_total_frames));
	if (!"""+input_vars[2]+""") {
		particle_frame = clamp(particle_frame, 0.0, particle_total_frames - 1.0);
	} else {
		particle_frame = mod(particle_frame, particle_total_frames);
	}	flipbook_uv /= vec2(h_frames, v_frames);
	flipbook_uv += vec2(mod(particle_frame, h_frames) / h_frames, floor((particle_frame + 0.5) / h_frames) / v_frames);
	"""+output_vars[0]+"""= flipbook_uv;
	"""
	return ""
```

**Comment by Chaosus** [comment]
I've just added a post-check for a string containing "null" - I've didn't found another way to prevent this - since `_get_func_code` is intended to return a string.

**Review by QbieShay — APPROVED** [review]
Tested with the script linked in the isssue comments.

This feature unlocks work on higher level nodes.

**Review by akien-mga — APPROVED** [review]
The documentation looks good :+1: 

@reduz wanted to have a look at the proposal and PR before merging.

**Comment by akien-mga** [comment]
Discussed in PR review meeting, it's good to merge after rebase :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
